app/services/groups/add_users_service.rb

```diff
@@ -96,25 +80,56 @@ def add_to_user_and_projects_cte
+          AND members.project_id = group_memberships.project_id
+          AND members.id IS NOT NULL
+        ),
         -- insert the group user into members
         new_members AS (
-          INSERT INTO #{Member.table_name} (project_id, user_id)
-          SELECT group_memberships.project_id, found_users.user_id
+          INSERT INTO #{Member.table_name} (project_id, user_id, updated_at, created_at)
+          SELECT group_memberships.project_id, found_users.user_id, (SELECT time from timestamp), (SELECT time from timestamp)
           FROM found_users, group_memberships
-          -- We need to return all members for the given group memberships
-          -- even if they already exist as members (e.g., added individually) to ensure we add all roles
-          -- to mark that we reset the created_at date since replacing the member
-          ON CONFLICT(project_id, user_id) DO UPDATE SET updated_at = CURRENT_TIMESTAMP
+          WHERE NOT EXISTS (SELECT 1 FROM existing_members WHERE existing_members.user_id = found_users.user_id AND existing_members.project_id = group_memberships.project_id)
+          ON CONFLICT(project_id, user_id) DO NOTHING
           RETURNING id, user_id, project_id
-        )
+        ),
         -- copy the member roles of the group
-        INSERT INTO #{MemberRole.table_name} (member_id, role_id, inherited_from)
-        SELECT new_members.id, group_roles.role_id, group_roles.member_role_id
-        FROM group_roles
-        JOIN new_members ON group_roles.project_id = new_members.project_id
-        -- Ignore if the role was already inserted by us
-        ON CONFLICT DO NOTHING
+        add_roles AS (
+          INSERT INTO #{MemberRole.table_name} (member_id, role_id, inherited_from)
+          SELECT members.id, group_roles.role_id, group_roles.member_role_id
+          FROM group_roles
+          JOIN
+            (SELECT * FROM new_members UNION SELECT * from existing_members) members ON group_roles.project_id = members.project_id
+          -- Ignore if the role was already inserted by us
+          ON CONFLICT DO NOTHING
+          RETURNING id, member_id, role_id
+        ),
+        -- get the ids of members where roles have been added the member did not have before
+        members_with_added_roles AS (
+          SELECT DISTINCT add_roles.member_id
+          FROM add_roles
+          WHERE NOT EXISTS
+            (SELECT 1 FROM #{MemberRole.table_name}
+              WHERE #{MemberRole.table_name}.member_id = add_roles.member_id
+              AND #{MemberRole.table_name}.role_id = add_roles.role_id
+              AND #{MemberRole.table_name}.id != add_roles.id)
+        ),
+        touch_existing_members AS (
+          UPDATE members SET updated_AT = CURRENT_TIMESTAMP
+          WHERE id IN (SELECT id from existing_members)
+          AND id IN (SELECT member_id from members_with_added_roles)
+        )
+
+        SELECT member_id from members_with_added_roles
```

Comment:
Now I see where that diff ended up

Reply:
Yes, there was quite some code added. I hope (and I currently honestly don't know) that it will turn out to be easier to maintain.